api/oci/ociutils/ref.go

```diff
@@ -0,0 +1,83 @@
+package ociutils
+
+import (
+	"strings"
+
+	"github.com/mandelsoft/goutils/generics"
+	"github.com/opencontainers/go-digest"
+)
+
+// ParseVersion parses the version part of an OCI reference consisting
+// of an optional tag and/or digest.
+func ParseVersion(vers string) (ArtVersion, error) {
+	if strings.HasPrefix(vers, "@") {
+		dig, err := digest.Parse(vers[1:])
+		if err != nil {
+			return ArtVersion{}, err
+		}
+		return ArtVersion{
+			Tag:    nil,
+			Digest: &dig,
+		}, nil
+	}
+
+	i := strings.Index(vers, "@")
+	if i > 0 {
+		dig, err := digest.Parse(vers[i+1:])
+		if err != nil {
+			return ArtVersion{}, err
+		}
+		return ArtVersion{
+			Tag:    generics.Pointer(vers[:i]),
+			Digest: &dig,
+		}, nil
+	}
+	return ArtVersion{
+		Tag: &vers,
+	}, nil
+}
+
+// ArtVersion is the version part of an OCI reference consisting of an
+// optional tag and/or digest. Both parts may be nil, if a reference
+// does not include a version part.
+type ArtVersion struct {
```

Comment:
The reason is, that this is a type extracted from the already existing type oci.ArtSpec. It had to be moved out of the oci package to avoid a package usage cycle when using this extracted type.,